### PR TITLE
Request for feedback: add initial audit log retrieval support

### PR DESCRIPTION
**Pull request by jchristgit**
This PR adds support for retrieving audit logs. Right now it's more or less just parsing the response from the API. However, the API is a mess. A slight mess. A big mess. Pick your poison.

Since I'm not sure if we should structure this nicer (e.g. should we parse each audit log event type into its own struct? what about new audit log events here). Would be happy for some feedback.

### CODE REVIEW

File: lib/nostrum/struct/guild/audit_log.ex [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  @type t :: %__MODULE__{
+          entries: entries,
+          users: users,
+          webhooks: webhooks
+        }
+
+  @doc false
+  def to_struct(map) do
+    struct(
+      __MODULE__,
```
Comment (Kraigie) [inline lib/nostrum/struct/guild/audit_log.ex]: Any particular reason you're doing the updates this way? Discord isn't exactly known to be super faithful about the fields they send. I'd suggest you check out how any other struct modules does this and see if it'd be better to do it similarly here, either just to be more consistent, or to handle any possible missing keys.

File: lib/nostrum/struct/guild/audit_log_entry.ex [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    changes: changes,
+    id: id,
+    options: options,
+    reason: reason,
+    target_id: target_id,
+    user_id: user_id
+  }
+
+  @doc false
+  def to_struct(map) do
```
Comment (Kraigie) [inline lib/nostrum/struct/guild/audit_log_entry.ex]: Similarly as above. I think we should try to be consistent with our method of transferring these maps to structs.

File: lib/nostrum/struct/guild/audit_log.ex [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  Represents a guild's audit log.
+  """
+
+  alias Nostrum.Struct.Guild.AuditLogEntry
+  alias Nostrum.Struct.{User, Webhook}
+  alias Nostrum.Util
+
+  defstruct [:entries, :users, :webhooks]
+
+  @typedoc "Entries of this guild's audit log."
```
Comment (Kraigie) [inline lib/nostrum/struct/guild/audit_log.ex]: We don't typically have periods at the end of the typedoc definitions

**Comment by Kraigie** [comment]
I don't think we should make structs for each type of event. They can match on the audit log event if they want to. It'd be a mess on our end to support all of that.

I'll take another look at the actual implementation later to make sure the types and such are OK.

Also, don't forget to run the formatter. 🙃

File: lib/nostrum/struct/guild/audit_log.ex (hunk 1 above)
Comment (jchristgit) [reply lib/nostrum/struct/guild/audit_log.ex]: This seemed the most intuitive - I see the point about missing keys though, maybe I should handle that more gracefully. Also, I could just use `%__MODULE__` directly. For `audit_log_entries`, I'd have to rename the key in `Map.update`, which means temporarily saving the key, deleting the other key, and then putting in the new key, which I'm not a huge fan of. Should I go with that approach or stick with this one and be more graceful about missing keys?

File: lib/nostrum/struct/guild/audit_log.ex (hunk 3 above)
Comment (jchristgit) [reply lib/nostrum/struct/guild/audit_log.ex]: Removed in 44473a6

**Comment by Kraigie** [comment]
Thanks!

**Pull request merged**